### PR TITLE
Trace fs operations

### PR DESCRIPTION
**Pull request by kislaykishore**
This adds support for tracing fs operations.

### Description
Perf results
image

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/fs/wrappers/monitoring.go [hunk 1]
```diff
@@ -327,9 +329,17 @@ func (fs *monitoring) Destroy() {
 type wrappedCall func() error
 
 func invokeWrapped(ctx context.Context, opName string, w wrappedCall) error {
+	// SpanKind is server since Fuse is a server for Kernel requests.
```
Comment (charith87) [inline internal/fs/wrappers/monitoring.go:332]: I think we can have this flag protected. Considering this is going to get invoked for every fs call. 
Is the sampling set in the client code or in the exporter or in cloud tracer. 
Would the flag protection be added after config cli changes merged?

Comment (charith87) [reply internal/fs/wrappers/monitoring.go:332]: Discussed offline. Without the TracerProvider its a no-op. 
Discussed to do execute-perf-test and share results.

File: internal/monitor/trace.go [hunk 2]
```diff
@@ -16,15 +16,23 @@ package monitor
...
+var tracerProvider trace.TracerProvider = noop.NewTracerProvider()
+var once sync.Once
+
 // StartSpan creates a new span and returns it along with the context that's augmented with the newly-created span..
 func StartSpan(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
-	return otel.GetTracerProvider().Tracer(name).Start(ctx, spanName, opts...)
+	once.Do(func() {
+		tracerProvider = otel.GetTracerProvider()
+	})
+	return tracerProvider.Tracer(name).Start(ctx, spanName, opts...)
```
Comment (charith87) [inline internal/monitor/trace.go:37]: could we put the traceProvider.Tracer(name) also in the once.Do() section that we if its initializing something , its only done once.

File: internal/fs/server.go [hunk 3]
```diff
@@ -31,6 +32,6 @@ func NewServer(ctx context.Context, cfg *ServerConfig) (fuse.Server, error) {
 	}
 
 	fs = wrappers.WithErrorMapping(fs)
-	fs = wrappers.WithMonitoring(fs)
+	fs = wrappers.WithMonitoring(fs, otel.Tracer("fs"))
```
Comment (charith87) [inline internal/fs/server.go:35]: Should fs be cloud.google.com/gcsfuse here

File: internal/monitor/traceexporter.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	tp, shutdown, err := newTraceProvider(ctx, c)
+	if err == nil && tp != nil {
+		otel.SetTracerProvider(tp)
+		return shutdown
+	}
+	return nil
+}
+
+func newTraceProvider(ctx context.Context, c *cfg.Config) (trace.TracerProvider, ShutdownFn, error) {
+	switch c.Monitoring.ExperimentalTracingMode {
```
Comment (charith87) [inline internal/monitor/traceexporter.go:45]: Should there be a case here where ExperimentalTracingMode is not defined
Or c.Monitoring itself null.

Comment (kislaykishore) [reply internal/monitor/traceexporter.go:45]: c.Monitoring cannot be nil. It'll be empty but never nil. Same goes for ExperimentalTracingMode as well.

**Review by charith87 — APPROVED** [review]
Overall LGTM. As its still WIP and tests would eventually be added as and when implemented for the new flags and exporters.

File: internal/monitor/traceexporter.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		resource.WithDetectors(gcp.NewDetector()),
+		resource.WithTelemetrySDK(),
+		resource.WithAttributes(
+			semconv.ServiceName(appName),
+		),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	tp := sdktrace.NewTracerProvider(sdktrace.WithBatcher(exporter), sdktrace.WithResource(res), sdktrace.WithSampler(sdktrace.TraceIDRatioBased(c.Monitoring.ExperimentalTracingSamplingRatio)))
```
Comment (vadlakondaswetha) [inline internal/monitor/traceexporter.go:85]: nit: i would suggest adding a new line after if clause to clear tell that if is ended here. improves readability

File: cfg/config.go [hunk 6]
```diff
@@ -415,6 +419,26 @@ func BindFlags(v *viper.Viper, flagSet *pflag.FlagSet) error {
 		return err
 	}
 
+	flagSet.StringP("experimental-tracing-mode", "", "", "Experimental: specify tracing mode")
+
+	if err := flagSet.MarkDeprecated("experimental-tracing-mode", "Experimental flag: could be dropped even in a minor release."); err != nil {
```
[the remark this answers is not in the document]
Comment (kislaykishore) [reply cfg/config.go:424]: The intention is to hide the flag. I've marked it as hidden now.

File: internal/monitor/traceexporter.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+type ShutdownFn func(ctx context.Context) error
+
+// SetupTracing bootstraps the OpenTelemetry tracing pipeline.
+func SetupTracing(ctx context.Context, c *cfg.Config) ShutdownFn {
+	tp, shutdown, err := newTraceProvider(ctx, c)
+	if err == nil && tp != nil {
+		otel.SetTracerProvider(tp)
+		return shutdown
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (kislaykishore) [reply internal/monitor/traceexporter.go:47]: Done

File: internal/fs/server.go [hunk 8]
```diff
@@ -20,17 +20,20 @@ import (
...
+// NewServer creates a fuse file system server according to the supplied configuration.
 func NewServer(ctx context.Context, cfg *ServerConfig) (fuse.Server, error) {
 	fs, err := NewFileSystem(ctx, cfg)
 	if err != nil {
 		return nil, fmt.Errorf("create file system: %w", err)
 	}
 
 	fs = wrappers.WithErrorMapping(fs)
-	fs = wrappers.WithMonitoring(fs)
+	fs = wrappers.WithMonitoring(fs, otel.Tracer(name))
```
[the remark this answers is not in the document]
Comment (kislaykishore) [reply internal/fs/server.go]: Makes sense.

File: internal/fs/wrappers/monitoring.go [hunk 9]
```diff
@@ -326,126 +330,134 @@ func (fs *monitoring) Destroy() {
 
 type wrappedCall func() error
 
-func invokeWrapped(ctx context.Context, opName string, w wrappedCall) error {
+func (fs *monitoring) invokeWrapped(ctx context.Context, opName string, w wrappedCall) error {
+	// SpanKind is server since Fuse is a server for Kernel requests.
```
[the remark this answers is not in the document]
Comment (kislaykishore) [reply internal/fs/wrappers/monitoring.go]: Let me know if it looks better now.

**Review by vadlakondaswetha — APPROVED** [review]

**Pull request merged**